app/src/main/java/co/runloop/influencer/fragment/ContactsFragment.java

```diff
@@ -0,0 +1,110 @@
+import android.util.Log;
+import android.view.LayoutInflater;
+import android.view.View;
+import android.view.ViewGroup;
+
+import java.util.List;
+
+import co.runloop.influencer.R;
+import co.runloop.influencer.adapter.ContactsAdapter;
+import co.runloop.influencer.data.ContactsProvider;
+import co.runloop.influencer.model.Contact;
+import co.runloop.influencer.viewmodel.ContactsViewModel;
+
+public class ContactsFragment extends BaseFragment {
+
+    private static final String TAG = ContactsFragment.class.getSimpleName();
+
+    private static final int READ_CONTACT_PERM_RC = 101;
+
+    public static ContactsFragment newInstance() {
+        Bundle args = new Bundle();
+        ContactsFragment fragment = new ContactsFragment();
+        fragment.setArguments(args);
+
+        return fragment;
+    }
+
+    private RecyclerView recyclerView;
+    private ContactsAdapter adapter;
+    private ContactsViewModel contactsViewModel;
+
+    @Override
+    public void onCreate(@Nullable Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        adapter = new ContactsAdapter();
+        if (isActionBarAvailable()) {
+            getBaseActivity().getSupportActionBar().setTitle(R.string.contacts);
+        }
+    }
+
+    @Nullable
+    @Override
+    public View onCreateView(@NonNull LayoutInflater inflater,
+                             @Nullable ViewGroup container,
+                             @Nullable Bundle savedInstanceState) {
+        View root = inflater.inflate(R.layout.frag_contacts_list, container, false);
+
+        recyclerView = root.findViewById(R.id.frag_contacts_list_recyclerview);
+        recyclerView.setLayoutManager(new LinearLayoutManager(getActivity()));
+        recyclerView.setAdapter(adapter);
+
+        contactsViewModel = ViewModelProviders
+                .of(this)
+                .get(ContactsViewModel.class);
+        contactsViewModel.getContacts().observe(this, (@Nullable List<Contact> contacts) -> {
+            adapter.setContacts(contacts);
+            adapter.notifyDataSetChanged();
+        });
+        if (contactsIsAvailable()) {
+            contactsViewModel.onPermissionGranted();
+            if(contactsViewModel.getContacts().getValue() == null) {
+                contactsViewModel.loadAll();
+            }
+        }
+
+        return root;
+    }
+
+    @Override
+    public void onRequestPermissionsResult(int requestCode,
+                                           @NonNull String[] permissions,
+                                           @NonNull int[] grantResults) {
+        if (contactsIsAvailable()) {
+            contactsViewModel.onPermissionGranted();
+            contactsViewModel.loadAll();
+        }
+    }
+
+    @Override
+    public void onResume() {
+        super.onResume();
+        if (!contactsIsAvailable()) {
+            requestReadContactsPermission();
+        }
+    }
+
+    private boolean contactsIsAvailable() {
```

Comment:
This will be very heavy implementation, and it means keep links to Fragment and Activity.
Because they have different methods to request and handle permissions

Reply:
Well, this query method would be the same any where you would put it. It could (and maybe should?) be static... (and take a context parameter)

I was hoping there is a way to ask for permissions via a context and object and not a direct reference to an activity or fragment, but maybe it's impossible... 

The same approach with accepting a parameter could be taken, and then you could still ask the `provider` to do this for you, and not have to copy/paste this code every time you need to work with contacts

last point - you pass `getActivity` to the permission request. Why not `getContext` ? (either way, maybe a null check is also a good idea)

Reply:
last point - you pass getActivity to the permission request. Why not getContext ? (either way, maybe a null check is also a good idea)
-> getActivity is very default option for me, it cover almost all needed context options. Is there some hidden problems in use getActivity ?

Reply:
Fragment cannot exists without activity, why we need to check null ? Of Course we can try to call it before onAttach, but in this case we work in lifecycle

Reply:
Yes, I understand that `getActivity` is kinda default option, not only for you. But.. have you read the article I referenced to? Seems there's a difference between UI Context and non-UI context... And also, there can be a difference in the life-span of certain context, and we want to make sure we don't supply a short lived context to a long living entity...

So, I'm saying that we need to consider these things, and not just use the "default" way we have been using so far. 

Regarding the lifecycle of the fragment, it's not 100% true that the fragment cannot live without the activity. It's mostly true, and it may be true for the use case of the app, but again, I want us to try and decide on practices we follow.... ways that will enable us to reduce bugs and crashes to a minimum at an early stage.... Can we be absolutely certain that the context cannot be null in this case? If we can, then ok... If not, it's probably worth checking :)